pkg/platform/kube/provisioner/apigateway/provisioner.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (p *Provisioner) CreateOrUpdateAPIGateway(ctx context.Context, apiGateway *nuclioio.NuclioAPIGateway) error {
+	var appliedIngressNames []string
+
+	apiGateway.Status.Name = apiGateway.Spec.Name
+
+	if err := p.validateSpec(apiGateway); err != nil {
+		return errors.Wrap(err, "Api gateway spec validation failed")
+	}
+
+	// generate an ingress for each upstream
+	upstreams := apiGateway.Spec.Upstreams
+	ingresses := map[string]*ingress.Resources{}
+
+	// always try to remove previous canary ingress first, because
+	// nginx returns 503 on all requests if primary service == secondary service. (happens on every promotion)
+	// so during promotion all requests will be sent to the primary ingress
+	p.tryRemovePreviousCanaryIngress(ctx, apiGateway)
+
+	if len(upstreams) == 1 {
+
+		// create just a single ingress
+		ingressResources, err := p.generateNginxIngress(ctx, apiGateway, upstreams[0])
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate nginx ingress")
+		}
+
+		ingresses[ingressResources.Ingress.Name] = ingressResources
+
+	} else if len(upstreams) == 2 {
+		var canaryUpstream platform.APIGatewayUpstreamSpec
+		var baseUpstream platform.APIGatewayUpstreamSpec
+
+		// determine which upstream is the canary one
+		if upstreams[0].Percentage != 0 {
+			baseUpstream = upstreams[1]
+			canaryUpstream = upstreams[0]
+		} else if upstreams[1].Percentage != 0 {
+			baseUpstream = upstreams[0]
+			canaryUpstream = upstreams[1]
+		} else {
+			return errors.New("Percentage must be set on one of the upstreams (canary)")
+		}
+
+		// validity check
+		if canaryUpstream.Percentage > 100 || canaryUpstream.Percentage < 1 {
+			return errors.New("The canary upstream percentage must be between 1 and 100")
+		}
+
+		// add the base ingress
+		baseIngressResources, err := p.generateNginxIngress(ctx, apiGateway, baseUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the base nginx ingress")
+		}
+		ingresses[baseIngressResources.Ingress.Name] = baseIngressResources
+
+		// add the canary ingress
+		canaryIngressResources, err := p.generateNginxIngress(ctx, apiGateway, canaryUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the canary nginx ingress")
+		}
+		ingresses[canaryIngressResources.Ingress.Name] = canaryIngressResources
+	}
+
+	// create ingresses
+	for ingressName, ingressResources := range ingresses {
+		if _, _, err := p.ingressManager.CreateOrUpdateIngressResources(ingressResources); err != nil {
+			p.Logger.WarnWithCtx(ctx, "Failed to create/update api-gateway ingress resources",
+				"err", errors.Cause(err),
+				"ingressName", ingressName,
+				"appliedIngressNames", appliedIngressNames)
+			return errors.New("Failed to create/update api-gateway ingress resources")
+		}
+
+		appliedIngressNames = append(appliedIngressNames, ingressName)
+	}
+
+	// sleep 4 seconds as a safety, so nginx will finish updating the ingresses properly (it takes time)
```

Comment:
is there a programatic way to do it? `retryUntilSuccessfull(some-condition)`?

Reply:
none that I could come up with..it's an internal nginx controller delay that we cannot check when it passes

Reply:
I agree this is very icky (though I know why it's here).
We should reconsider whether we want it and if not just remove this (and deal with the consequences internally)
Currently it's not super harmful and may even help with stability for testing this, if we find in nuclio tests that it's not needed, then this is harming performance and we'll remove it